### ISSUE TITLE
"flutter run" should support ^R for a cold restart

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#77415]
Watching people use "flutter run", one thing I see happen a lot is that they will add a new plugin dependency, do a hot reload, it won't work, do a hot restart, it won't work, then hit ^C and rerun flutter run (to get the plugin compiled in).

It would be neat if you could just hit ^R to do a cold restart without quitting the tool.

**Comment by jonahwilliams** [comment]
Here is a proof of concept: https://github.com/flutter/flutter/pull/77418 , this only works for android with dds disabled.

@bkonyi  , when running with dds I'm not able to restart it. is there a good way I could keep dds running (to preserve the vm service URI), but detach and re-attach a Dart VM

**Comment by jonahwilliams** [comment]
this may also require some additional affordances for IDEs, since they'll need to be able to discover the new VM Service URI

**Comment by bkonyi** [comment]
> Here is a proof of concept: #77418 , this only works for android with dds disabled.
> 
> @bkonyi , when running with dds I'm not able to restart it. is there a good way I could keep dds running (to preserve the vm service URI), but detach and re-attach a Dart VM

We want to keep DDS alive so clients don't get dropped after a cold restart I take it? If that's not a concern, we can just kill DDS and relaunch with the same host/port, although we'll need to do something about the authentication codes in that situation.

Otherwise, there's currently no way to switch the target of a DDS instance. Since this case would effectively be connecting to a new target application, we'd just have to reset state and reconnect within DDS without killing connections to its clients, so I don't think that would be very difficult to add.

**Comment by zanderso** [comment]
@Hixie I'm a bit concerned that if people aren't reading that they can quit the tool with `q` (and instead doing `^C`), then they also might not read that they can do a cold restart with some other key combo.

**Comment by Hixie** [comment]
`^C` does the same thing as `q` as far as I can tell?

**Comment by jonahwilliams** [comment]
> We want to keep DDS alive so clients don't get dropped after a cold restart I take it? If that's not a concern, we can just kill DDS and relaunch with the same host/port, although we'll need to do something about the authentication codes in that situation.

Yeah, Ideally devtools/IDEs would keep working even if the state got cleared. What is the issue with auth codes?

**Comment by zanderso** [comment]
Yes, we had to make `^C` do the same thing after noticing that many people do that instead of `q`. Luckily, `^C` on Windows sends `SIGINT`, which is one of the few signals that we can handle there: https://api.dart.dev/dev/2.13.0-109.0.dev/dart-io/ProcessSignal/watch.html. (Paradoxically, on Windows, Dart code can only *send* `SIGTERM`: https://api.dart.dev/dev/2.13.0-109.0.dev/dart-io/Process/killPid.html)

**Comment by jonahwilliams** [comment]
I would rather keep this as a terminal key for discoverability reasons.

**Comment by jonahwilliams** [comment]
I managed to get dds working anyway. The entire process probably tears down more than it should - closing devtools and whatnot. It would be better if we could keep that open

**Comment by matanlurey** [comment]
@bkonyi Could use your 👁️

**Comment by bkonyi** [comment]
While this would be interesting, I don't think it adds enough value to the developer experience to warrant implementing, wiring up in the IDEs, etc.
